### ISSUE TITLE
Document doesn't appear in results even though all tokens have a match

### ISSUE THREAD

**Issue by matteocontrini** [typesense/typesense#562]
## Description

I'm trying out Typesense on a dataset I've built with about 30k discussions from an online forum. The relevant fields for each indexed items are the `title` (string) and an array of `tags` (strings, faceted).

I've noticed that if a search query is composed of the title of a document with a "tag" in the middle, the document is not matched. If I move the "tag" at the end of the query, the document is returned.

I know it's hard to understand and I couldn't get to a minimum reproducible example (I'm seeing the issue only with a lot of data in the collection), so I'll try with an example.

## Steps to reproduce

I have the following indexed item (this is Italian):

- title = `Rallentamento alla sera (Torino)`
- tags = [`FTTH`, `Open Fiber`, `Speedtest`]

I run the following query, where there are two words of the title ("rallentamento" and "torino") with a tag in the middle ("open fiber"):

```
{
  "q": "rallentamento open fiber torino",
  "query_by": "title,tags",
  "per_page": 50
}
```

## Actual Behavior

The item **doesn't** appear in the first 50 results. Instead, I see other documents where it **never** happens that all the 4 tokens are matched (at most 3). These results don't seem relevant to me because they all miss an important word of the query.

If I instead run this query, where the tag "open fiber" is moved at the end of the query:

```
{
  "q": "rallentamento torino open fiber",
  "query_by": "title,tags",
  "per_page": 50
}
```

The first result is the item above.

```
{
    "document": {
        "id": "24734",
        "tags": [
            "FTTH",
            "Open Fiber",
            "Speedtest"
        ],
        "title": "Rallentamento alla sera (Torino)"
    },
    "highlights": [
        {
            "field": "tags",
            "indices": [
                1
            ],
            "matched_tokens": [
                [
                    "Open",
                    "Fiber"
                ]
            ],
            "snippets": [
                "<mark>Open</mark> <mark>Fiber</mark>"
            ]
        },
        {
            "field": "title",
            "matched_tokens": [
                "Rallentamento",
                "Torino)"
            ],
            "snippet": "<mark>Rallentamento</mark> alla sera (<mark>Torino)</mark>"
        }
    ],
    "text_match": 4419505028190
}
```

## Expected Behavior

I expect to see the document in the search results for `rallentamento open fiber torino`, because *every* token of the query appears at least once in the document fields.

## Metadata

**Typsense Version**: 0.22.2

**Comment by merabtenei** [comment]
Can you try with exhaustive_search: true,

```
{
  "q": "rallentamento torino open fiber",
  "query_by": "title,tags",
  "exhaustive_search": true,
  "per_page": 50
}
```

**Comment by matteocontrini** [comment]
If I try the query `rallentamento torino open fiber` (which would normally return the expected document as the first result) with exhaustive search the document doesn't appear anymore in the first 50 results.

If I try `rallentamento open fiber torino` (which is the problematic query) it doesn't change the result, the document doesn't show up.

**Comment by kishorenc** [comment]
@matteocontrini 

Typesense maintains indices at a field level, and does an `AND` search of words in the query. If a field does not contain all the words in the query, we then drop words in the query from right to left (or left to right) until we find results. 

One drawback of this approach to dropping words is that ordering matters but this is generally fine because people tend to write queries in field-matching order (e.g. `nike running shoes` (brand + category) or `restaurants new york` (place + city) etc.). 

If this assumption does not hold good for your application, I recommend adding a field which simply contains all words in the other fields as a "concatenated" text field, and use that as the last field in the `query_by` list. This way, you guarantee to accurately locate documents containing all query words and the the word ordering issue will not surface here.

**Comment by kishorenc** [comment]
Having said, I agree that this is still inconvenient, so we are planning to address this shortcoming in the next release.

**Comment by matteocontrini** [comment]
That makes sense, thanks for the explanation. Looking forward to improvements, thanks!

**Comment by kishorenc** [comment]
@matteocontrini 

I've published a new Docker build `0.23.0.rc53` that address the cross field matching issue that you noticed. Can you please try that out and let me know?

**Comment by toddlt** [comment]
> 

may I change the AND to OR by adding specific query param? thanks
